src/frequenz/channels/_base_classes.py

```diff
@@ -133,6 +40,11 @@ async def __anext__(self) -> T:
             StopAsyncIteration: if the receiver stopped producing messages.
             ReceiverError: if there is some problem with the receiver.
         """
+        # Break cyclic import
+        from ._exceptions import (  # pylint: disable=import-outside-toplevel,cyclic-import
+            ReceiverStoppedError,
+        )
```

Comment:
does this have a perf cost,  importing from inside `__anext__`?

It looks like the cyclic importing is necessary just because of type hint dependencies in `_exceptions.py`.  It might be interesting to examine switching to automatically generated typing stubs,  then types become independent:  https://mypy.readthedocs.io/en/stable/stubgen.html

It is hard for me to see the benefit in pulling the exceptions into a separate package,  when we're having to work around ourselves when importing the package.

Reply:
> does this have a perf cost

Good question, my guess is it shouldn't be too much, as import should first check if the module was already imported, so it should be just an extra lookup in the modules dict, which compared to sync primitives should be negligible, but I agree is probably the worse place to break a circular dependency.

I don't think generating stubs is a good idea, it was more of a hack for having at least *some* type info for libs that didn't use native type hints, but if you look at the stubgen example in the link you sent, it doesn't even infer the type of the method arguments.

> Stubgen generates draft stubs. The auto-generated stub files often require some manual updates, and most types will default to `Any`.

> It is hard for me to see the benefit in pulling the exceptions into a separate package, when we're having to work around ourselves when importing the package.

Yeah, for me is just an organizational improvement, is easier to find the code and to avoid noise when looking at unrelated stuff, but I do agree that it might not be the best approach if it needs *workarounding*.

I'll see if there is a better way to avoid the dependency loops and if not I might just move the exceptions back to the base classes module.

Reply:
OK, I found an *elegant* (by Python type hints standards :trollface:) way to break the circular dependency: [`typing.TYPE_CHECKING`](https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING).